### PR TITLE
[GLUTEN-4903][CELEBORN] Support multiple versions of Celeborn

### PR DESCRIPTION
**Pull request by kerwin-zk**
## What changes were proposed in this pull request?

Support multiple versions of Celeborn

#4903 

## How was this patch tested?

CI

### CODE REVIEW

**Comment by ulysses-you** [comment]
Is it possible to use one module to be compatible with multi-version celeborn ? It is hard to maintain if we always create new module for new coming celeborn version. Besides, it's hard to work if someone use 0.3 branch but backport some 0.4 patch..

**Comment by kerwin-zk** [comment]
> Is it possible to use one module to be compatible with multi-version celeborn ? It is hard to maintain if we always create new module for new coming celeborn version. Besides, it's hard to work if someone use 0.3 branch but backport some 0.4 patch..
> 
> cc @pan3793

@ulysses-you I do agree that we need to consider the case of  ```it's hard to work if someone uses the 0.3 branch but backports some 0.4 patch```, and I will consolidate it into one.

File: .github/workflows/velox_be.yml [hunk 1]
```diff
@@ -450,11 +450,27 @@ jobs:
             --local --preset=velox --benchmark-type=h --error-on-memleak --off-heap-size=10g -s=1.0 --threads=16 --iterations=1 \
           && GLUTEN_IT_JVM_ARGS=-Xmx20G sbin/gluten-it.sh queries-compare \
             --local --preset=velox --benchmark-type=ds --error-on-memleak --off-heap-size=40g -s=10.0 --threads=32 --iterations=1'
-      - name: TPC-H SF1.0 && TPC-DS SF10.0 Parquet local spark3.2 with Celeborn
+      - name: TPC-H SF1.0 && TPC-DS SF10.0 Parquet local spark3.2 with Celeborn 0.4.0
         run: |
           $PATH_TO_GLUTEN_TE/$OS_IMAGE_NAME/gha/gha-checkout/exec.sh \
-          'wget https://archive.apache.org/dist/incubator/celeborn/celeborn-0.3.0-incubating/apache-celeborn-0.3.0-incubating-bin.tgz && \
-          tar xzf apache-celeborn-0.3.0-incubating-bin.tgz -C /opt/ && mv /opt/apache-celeborn-0.3.0-incubating-bin /opt/celeborn && cd /opt/celeborn && \
+          'wget https://archive.apache.org/dist/incubator/celeborn/celeborn-0.4.0-incubating/apache-celeborn-0.4.0-incubating-bin.tgz && \
```
Comment (pan3793) [inline .github/workflows/velox_be.yml:456]: CI only covers 0.3.2 and 0.4.0, what about 0.3.0 and 0.3.1? we actually have breaking changes on the APIs that Gluten used in the 0.3.x serial

Comment (kerwin-zk) [reply .github/workflows/velox_be.yml:456]: @pan3793  CI of 0.3.0 and 0.3.1 has been successfully run. Mainly considering the running time of CI, only one version of 0.3.x is currently retained. If everyone thinks that CI duration is not an issue, I can add it.

File: gluten-celeborn/pom.xml [hunk 2]
```diff
@@ -26,6 +26,16 @@
...
+      <exclusions>
+        <exclusion>
+          <groupId>org.apache.celeborn</groupId>
+          <artifactId>celeborn-client-spark-${spark.major.version}_${scala.binary.version}</artifactId>
+        </exclusion>
+        <exclusion>
+          <groupId>org.apache.celeborn</groupId>
+          <artifactId>celeborn-spark-${spark.major.version}-columnar-shuffle_${scala.binary.version}</artifactId>
+        </exclusion>
+      </exclusions>
```
Comment (pan3793) [inline gluten-celeborn/pom.xml:38]: the shaded artifacts should not expose the shipped transitive deps, seems there are issues in 0.4.0-incubating cc @cfmcgrady

Comment (kerwin-zk) [reply gluten-celeborn/pom.xml:38]: @pan3793  I am handling this issue.

**Review by pan3793 — APPROVED** [review]

File: gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 3]
```diff
@@ -284,24 +300,57 @@ public <K, V> ShuffleWriter<K, V> getWriter(
...
+        // for Celeborn 0.4.0
+        try {
+          Method celebornShuffleIdMethod =
+              SparkUtils.class.getMethod(
+                  "celebornShuffleId",
+                  ShuffleClient.class,
+                  CelebornShuffleHandle.class,
+                  TaskContext.class,
+                  boolean.class);
+          shuffleId = (int) celebornShuffleIdMethod.invoke(null, shuffleClient, h, context, true);
```
Comment (ulysses-you) [inline gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:335]: ditto

File: pom.xml [hunk 4]
```diff
@@ -47,7 +47,7 @@
     <sparkbundle.version>3.2</sparkbundle.version>
     <spark.version>3.4.2</spark.version>
     <sparkshim.artifactId>spark-sql-columnar-shims-spark32</sparkshim.artifactId>
-    <celeborn.version>0.3.0-incubating</celeborn.version>
+    <celeborn.version>0.3.2-incubating</celeborn.version>
```
Comment (ulysses-you) [inline pom.xml:50]: shall we add profile for 0.4 ?

File: gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 5]
```diff
@@ -284,24 +300,57 @@ public <K, V> ShuffleWriter<K, V> getWriter(
       ShuffleHandle handle, long mapId, TaskContext context, ShuffleWriteMetricsReporter metrics) {
     try {
       if (handle instanceof CelebornShuffleHandle) {
+        byte[] extension;
+        try {
+          Field field = CelebornShuffleHandle.class.getDeclaredField("extension");
```
Comment (ulysses-you) [inline gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:305]: ditto

File: gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 6]
```diff
@@ -99,6 +107,18 @@ public CelebornShuffleManager(SparkConf conf) {
     this.conf = conf;
     this.celebornConf = SparkUtils.fromSparkConf(conf);
     this.fallbackPolicyRunner = new CelebornShuffleFallbackPolicyRunner(celebornConf);
+
+    this.shuffleIdTracker =
+        CelebornUtils.createInstance(CelebornUtils.EXECUTOR_SHUFFLE_ID_TRACKER_NAME);
+    try {
+      Method clientFetchThrowsFetchFailureMethod =
+          celebornConf.getClass().getDeclaredMethod("clientFetchThrowsFetchFailure");
+      this.throwsFetchFailure = (Boolean) clientFetchThrowsFetchFailureMethod.invoke(celebornConf);
```
Comment (ulysses-you) [inline gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:116]: I think such reflect can be moved to `CelebornUtils`

File: gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 7]
```diff
@@ -190,27 +158,61 @@ private void initializeLifecycleManager() {
...
+                      Method unregisterAllMapOutputMethod =
+                          SparkUtils.class.getMethod(
+                              "unregisterAllMapOutput", MapOutputTrackerMaster.class, int.class);
+                      unregisterAllMapOutputMethod.invoke(null, mapOutputTracker, shuffleId);
+                    } catch (Exception e) {
+                      throw new RuntimeException(e);
+                    }
+                  };
+
+              registerShuffleTrackerCallbackMethod.invoke(lifecycleManager, consumer);
```
Comment (ulysses-you) [inline gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:185]: ditto

File: gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java [hunk 8]
```diff
@@ -220,6 +222,25 @@ public <K, V, C> ShuffleHandle registerShuffle(
...
     initializeLifecycleManager();
+
+    // for Celeborn 0.4.0
+    try {
+      Method registerAppShuffleDeterminateMethod =
+          lifecycleManager
+              .getClass()
+              .getDeclaredMethod("registerAppShuffleDeterminate", Integer.TYPE, Boolean.TYPE);
+
+      registerAppShuffleDeterminateMethod.invoke(
```
Comment (ulysses-you) [inline gluten-celeborn/common/src/main/java/org/apache/spark/shuffle/gluten/celeborn/CelebornShuffleManager.java:233]: ditto

**Review by ulysses-you — COMMENTED** [review]
thank you @kerwin-zk

File: pom.xml (hunk 4 above)
Comment (kerwin-zk) [reply pom.xml:50]: In theory, Celeborn only needs one version, as differences between versions have been eliminated through reflection.

**Review by ulysses-you — APPROVED** [review]

**Pull request merged**